### ISSUE TITLE
Feature request: option to not move the viewport when entering text

### ISSUE THREAD

**Issue by WalterGR** [Quicksaver/FindBar-Tweak#179]
FindBar Tweak is great - fixes a lot of annoyances I have with Firefox's built-in searching.

One annoyance it doesn't fix is that when you press Ctrl+F and start typing text to search, the viewport moves, completely losing where you were in the page.

Sometimes.

I'm trying to figure out Firefox's twisted logic, but I can't.  For example:

1.  Go to http://kb.mozillazine.org/Run_multiple_copies_of_Thunderbird_at_the_same_time
2.  Scroll to the bottom of the page.  If necessary, reduce the height of the browser window so that only the bottom half of the page (approximately) is visible.
3.  Press Ctrl+F and type "Thunder".  Matches will be highlighted but the viewport doesn't move, even though the first match is at the very top of the page, which isn't visible.
4.  Press Ctrl+F and type "instance".

Expected result:

The same behavior as when searching for "Thunder": matches will be highlighted but the viewport doesn't move, even though the first match is at the very top of the page, which isn't visible.

Actual result:

Firefox moves the viewport so that the first match is visible - even though there were matches for "instance" that were visible without moving the viewport.

I know this is a FF issue, but having an add-on that fixes it would make the search experience so much better.  Thanks for your consideration!

**Comment by Quicksaver** [comment]
I honestly never noticed firefox did this, it must be a new feature because I'm sure it didn't use to do this in the past.

In your steps you're missing something very important to understand this. When typing "thunder", it doesn't scroll to the first match because there are matches visible, and those take precedence (this is the new feature I was talking about, it didn't use to do this in the past). Whereas when you type "instance" there are no visible matches at the end of the document, so it has to wrap back up to the top. If you scroll the document to where there is a visible "instance" match and only then find it, it won't scroll to the top, it will start on that visible match.

There's two things about this worth mentioning.
- I don't consider this a bug at all, it makes sense that it starts finding from where you are, rather than always from the top. And of course if where you are doesn't have any matches, it scrolls to the first visible match. And if from where you are to the end of the document there are no matches (your "instance" case), it has to start from the top. This makes sense to me.
- Even if I wanted to do something about this, I'm not sure that I could. This appears to be something that happens within C++ code (think deep in the core of the browser), and those are completely inaccessible from my add-on.

I hope this clarifies what's happening at least. :)

**Comment by Quicksaver** [comment]
At least, for me that's what's happening and it is what I would expect it to do. If you're sure it is not starting from where it should (i.e. always scrolling to the top even when there are matches visible), you should open a bug in [bugzilla](https://bugzilla.mozilla.org/) and let them know.

**Comment by WalterGR** [comment]
I appreciate your detailed comment, but

> In your steps you're missing something very important to understand this. When typing "thunder", it doesn't scroll to the first match because there are matches visible, and those take precedence (this is the new feature I was talking about, it didn't use to do this in the past). Whereas when you type "instance" there are no visible matches at the end of the document, so it has to wrap back up to the top. 

I think you must have missed a few instances of "instance".  :)  There are matches for "instance" at the end of the document - several in fact, if your browser window's height has the viewport showing the bottom 50% of the page, as per the repro instructions.  There are 3 in the "New window by double clicking on folder" section and 1 in the "Two MinimizeToTray Plus icons in system tray" section.

Firefox doesn't need to move the viewport to find them.  The fact that the behavior is different for "Thunder" and "instance" is exactly why Firefox's default behavior is so frustrating.

> Even if I wanted to do something about this, I'm not sure that I could. This appears to be something that happens within C++ code (think deep in the core of the browser), and those are completely inaccessible from my add-on.

I'm a software engineer and I know C++.  Based on your extension, it seems like FF exposes an awful lot of the search functionality to extensions.  If I created my own extension, would I be able to fix this behavior for myself, or is really and truly in the C++ core?

Obviously if it's the latter I'll file a bug.

Thanks for your time.

**Comment by WalterGR** [comment]
Actually let me rephrase a couple things.  a. This is definitely a bug that I should file against FF.  b. My personal preference is that the viewport not move even if there aren't matches within it.  I understand that others may prefer something else.  My question about whether it's something an extension could address is out of curiosity, and not specifically related to FindBar-Tweak.

Thanks.

**Comment by Quicksaver** [comment]
>  There are matches for "instance" at the end of the document - several in fact, if your browser window's height has the viewport showing the bottom 50% of the page, as per the repro instructions.

This is what I meant, I'm sorry if I wasn't clear. If I resize the window to exactly half the height, "instance" is visible yeah, and if I search for it, the page doesn't scroll to the top, it finds those visible matches first. This is why it seems to me like you're experiencing a bigger issue and suggested you report it to mozilla. :)

>  My personal preference is that the viewport not move even if there aren't matches within it. I understand that others may prefer something else.

I think either you're missing something here, or I'm not fully understanding what you mean. There are actually two different types of operations going on when you perform a search in the page.
- "Find in Page": this finds the closest occurrence of your query in the page; also it's almost exactly the same operation as when you hit (Shift+)F3 to Find Previous/Next, except in these cases it just reuses a previous query. This is almost exclusively done in in C++ (through interface nsITypeAheadFind); basically the findbar just tells it "go to somewhere it has xyz" and the component finds it, selects it, scrolls the page and everything else.
- "Highlight all": this highlights all the occurrences of your query in the page. The actual finding of the matches is also done in C++ (through interface nsIFind), but in this case that interface just tells the findbar "here are your matches, do what you want with it". The findbar/add-on does the highlighting part, it just so happens that the add-on enables this by default and makes it seem like it's all the same thing. Firefox can also do this by itself but you have to press the "Highlight all" button every time. ;)

This is to say, I can understand that the page shouldn't scroll for the "Highlight all" (and it doesn't actually), but it still should scroll for the "Find in Page", only in case there are no visible matches where the page is currently of course (i.e. the bug). Now of course you can just stop it from doing the "Find in Page" operation completely, but I honestly think you'd be losing the whole point of the findbar...

> If I created my own extension, would I be able to fix this behavior for myself, or is really and truly in the C++ core?

Well this is a tough question for me as I only know enough C++ to barely follow it and know what it's doing. I know you can't modify the native components, but that's not to say you can't create your own, which I believe you can (but it's probably better you don't take my word for it on this).

This is all also entirely possible to be done in javascript (what most add-ons are written on, including FBT), I think. The "Highlight all" would probably be easier as a lot of it is javascript already. But to be honest, just thinking about how an entirely javascript-written "Find in Page" would work gives me a headache. Plus, I'm betting there would be a noticeable performance drop in doing this.

At least for now I'd suggest you report it to mozilla and see what they say (by the way, if you do let me know as I'd like to follow that bug as well). And if it starts to really bother you, you can start your research into fixing it yourself then. Unfortunately, like I said, I don't believe it is something I can do anything about in FBT; I'm sorry to disappoint you on this.
